### PR TITLE
rewrite ratelimiting and api route builder

### PR DESCRIPTION
**Pull request by devsnek**
**Please describe the changes this PR makes and why it should be merged:**
- rewrite api route builder to be faster/more effecient (also adds back methods e.g. `channels(id).messages(id)`)
- rewrite ratelimiting to be centralized around the actual ratelimit method and not the collecting of headers and stuff

**Semantic versioning classification:**  
- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by KingDGrizzle** [comment]
So, is it `[]` or is it `()`?
I've seen so many commits changing these its turning into a mystery

**Comment by devsnek** [comment]
@KingDGrizzle its actually whitespace now, we're switching to elixir

File: src/client/rest/RESTManager.js [hunk 1]
```diff
@@ -35,21 +34,17 @@ class RESTManager {
...
-        return SequentialRequestHandler;
-      case 'burst':
-        return BurstRequestHandler;
-      default:
-        throw new Error('RATELIMIT_INVALID_METHOD');
-    }
+    const method = this.client.options.apiRequestMethod;
+    if (typeof method === 'function') return method;
+    const handler = handlers[method];
+    if (!handler) throw new Error('RATELIMIT_INVALID_METHOD');
```
Comment (hydrabolt) [inline src/client/rest/RESTManager.js]: Makes more sense as INVALID_RATELIMIT_METHOD

Comment (devsnek) [inline src/client/rest/RESTManager.js]: they are named with category first, just look at the other ones

**Comment by meew0** [comment]
As this involves ratelimit code it should be tested extensively to ensure continued compliance

**Pull request merged**